File: lib/+acs/acs.m
function [ACS, preACS, preACS_unrestricted, dACS] = acs(GCSD2_avg, mapping, output_type)

    % Compute acoustic coherence state (ACS) quantities
    % based on averged secondary generalized cross spectral density (GCSD2)
    %
    % Input:
    %   GCSD2_avg: Averaged GCSD2 up until Nyq. Freq. Bin
    %   mapping: which mapping preACS->ACS to use. ("EXP")
    %   output_type: "cont" (continuous) or "bin" (binary)
    %               binary output is transformed via 
    %               thresholding (rules may differ depending on mapping)
    %
    % Output:
    %   ACS: Acoustic coherence state
    %   preACS: preliminary ACS, before mapping is applied
    %   preACS_unrestricted: preACS before temporal restrictions  are
    %                       applied
    %   dACS: differential ACS

    preACS_unrestricted = mean(abs(GCSD2_avg), 1); % no SFM
    preACS = preACS_unrestricted; %redundant
    dACS = 0; %redundant
    
    % Obtain ACS
    ACS = acs.acs__mapping(preACS_unrestricted, mapping, output_type);   

end

File: work/python/modules/acs.py
import numpy as np

class ACS:

    def __init__(self):
        self.preACS = 0
        self.ACS = 0
        self.mapping = 'EXP'

    def process(self, GCSD2_avg):
        # GCSD2_avg: exp. averaged GCSD2 up until nyquist-freq.
        self.preACS = np.mean(np.abs(GCSD2_avg), axis=0)
        self.ACS = self._map_EXP(self.preACS)
        return self.ACS

    def _map_EXP(self, preACS, binary=True):
        # exponential mapping in log-domain.
        a, b, c = -0.636, -3.362, 23.4
        basic_map = lambda x : a - b*np.exp(-c*x)
        chi_max = 0.25 # mapped to ACS=1
        maxLogSROEstErr = 0 # 1ppm
        out = basic_map(preACS) - maxLogSROEstErr
        out = out / np.abs(basic_map(chi_max))
        out = -out
        out[out < 0] = 0
        out[out > 1] = 1
        if binary:
            out[out > 0] = 1
        
        return out
